File: 2-trap1/trap1.c
#include <stdio.h>

/* Calculate local integral  */
double Trap(double left_endpt, double right_endpt, int trap_count, double base_len);

void main()
{
    /* Declaration */
    int n;
    double a;
    double b;
    double h;
    double local_int;
    double local_a;
    double local_b;

    /* Assignment */
    n = 1024;
    a = 0.0;
    b = 14.0;
    h = (b - a) / n;

    local_a = a + h;
    local_b = local_a + h;
    local_int = Trap(local_a, local_b, n, h);

    printf("With n = %d trapezoids, our estimate\n", n);
    printf("of the integral from %f to %f = %.15e\n", a, b, local_int);
}

/*------------------------------------------------------------------
 * Function:     Trap
 * Purpose:      Serial function for estimating a definite integral 
 *               using the trapezoidal rule
 * Input args:   left_endpt
 *               right_endpt
 *               trap_count 
 *               base_len
 * Return val:   Trapezoidal rule estimate of integral from
 *               left_endpt to right_endpt using trap_count
 *               trapezoids
 */
double Trap(double left_endpt, double right_endpt, int trap_count, double base_len)
{
    double estimate;
    double x;
    int i;

    estimate = (left_endpt * left_endpt + right_endpt * right_endpt) / 2.0;
    for (i = 1; i <= trap_count - 1; i++)
    {
        x = left_endpt + i * base_len;
        estimate += (x * x);
    }
    estimate = estimate * base_len;

    return estimate;
} /*  Trap  */

File: 2-trap1/trap1.f90
program trap1
    implicit none

    ! Calculate local integral
    INTERFACE 
    SUBROUTINE Trap(left_endpt, right_endpt, trap_count, base_len, estimate)
        REAL, INTENT(IN) :: left_endpt, right_endpt, base_len
        INTEGER, INTENT(IN) :: trap_count
        REAL, INTENT(OUT) :: estimate
    END SUBROUTINE Trap
    END INTERFACE

    ! Declaration of variables
    INTEGER :: n
    REAL :: a
    REAL :: b
    REAL :: h
    REAL :: local_int
    REAL :: local_a
    REAL :: local_b

    ! Assignment
    n = 1024
    a = 0.0
    b = 14.0
    h = (b - a) / n;

    local_a = a + h
    local_b = local_a + h

    CALL Trap (local_a, local_b, n, h, local_int)

    write (*, *) 'With n = ' , n, ' trapezoids, our estimate'
    write (*, *) 'of the integral from', a, ' to ' , b, ' = ', local_int

end program trap1

!---------------------------------------------------------------------
!
!  Subroutine for estimating a definite integral 
!  using the trapezoidal rule
! Input args:   left_endpt
!               right_endpt
!               trap_count 
!               base_len
! Output val:   Trapezoidal rule estimate of integral from
!               left_endpt to right_endpt using trap_count
!               trapezoids
!
!---------------------------------------------------------------------
SUBROUTINE Trap(left_endpt, right_endpt, trap_count, base_len, estimate)

    IMPLICIT NONE
    REAL, INTENT(IN) :: left_endpt
    REAL, INTENT(IN) :: right_endpt
    REAL, INTENT(IN) :: base_len
    INTEGER, INTENT(IN) :: trap_count
    REAL, INTENT(OUT) :: estimate
    REAL :: x
    INTEGER :: i

    estimate = ( left_endpt * left_endpt + right_endpt * right_endpt ) / 2.;

    do i = 1, trap_count, 1

        x = left_endpt + i * base_len
        estimate = estimate + ( x * x )

    end do

    estimate = estimate * base_len

END SUBROUTINE Trap